File: core/TemplateGeneratorNxM.py
import math
import numpy as np
from scipy.signal import find_peaks, wiener, peak_widths
from matplotlib import pyplot as plt

from ROOT import TGraph, TH1F, TLine

from utils import create_name_hist, get_angle_std, check_angle, vector_distribution, draw_hists, draw_graphs

class TemplateGeneratorNxM:

    def __init__( self, V, calc_r, calc_theta, E_min, E_max, vec_r_lim, mat_theta_lim ):

        # V is the cross-correlation function of the noise
        # calc_r and calc_theta are functors that calculate the radial and azimuthal partitions given the channel amplitudes
        # E_min and E_max define the bin in the total energy to which the templates correspond
        # Therefore, only events with total energy between E_min and E_max will be used to calculate the templates
        # vec_r_lim is a list that contains the limits of the bins in the radial partition
        # mat_theta_lim is a list of lists that, for each bin in the radial partition, contains the limits of the bins in the azimuthal partition
        # Note that, therefore, it is possible to define a segmentation such that the limits of the bins in the azimuthal partition depend on the
        # radial partition
        # As an example, see Fig. 12 of http://titus.stanford.edu/cdms_restricted/elias/2018_12_06-nxm_filter/index.html

        self.num_channels = len( V )
        self.num_bins_t = len( V[ 0 ][ 0 ] )

        self.loop_channels = range( self.num_channels )
        self.loop_bins_t = range( self.num_bins_t )

        self.J = [ [ V[ a ][ a ][ n ] for n in self.loop_bins_t ] for a in self.loop_channels ]

        self.calc_r = calc_r
        self.calc_theta = calc_theta

        self.E_min = E_min
        self.E_max = E_max

        # map_bins_part is the structure that, ultimately, relates each bin in the partition space with the corresponding template
        # It also holds the information about the bin limits provided by vec_r_lim and mat_theta_lim
        # Each element of map_bins_part corresponds to one bin in the radial partition
        # In turn, each such element contains a list where each element corresponds to one bin in the azimuthal partition
        # It is important to notice that the total number of bins registered in map_bins_part is larger than that defined by vec_r_lim and
        # mat_theta_lim
        # The reason is that, in order to handle general segmentations, each bin in the radial partition needs to be complemented by two
        # (or one, in some special cases) further bins that 1) are adjacent to it, and 2) have the same limits in the azimuthal partition as
        # it, called "sideband" bins
        # Note that these sideband bins overlap with other original bins, and therefore one event will, in general, contribute to built
        # templates in more than one bin

        self.map_bins_part = []
        self.cumul_S_fft = []
        self.event_counts = []

        for i in range( len( mat_theta_lim ) ):
            if i > 0:

                # This is the "bottom" sideband bin

                self.map_bins_part.append( ( vec_r_lim[ i-1 ], vec_r_lim[ i ], [] ) )

                for j in range( len( mat_theta_lim[ i ] ) ):
                    self.map_bins_part[ -1 ][ 2 ].append( ( mat_theta_lim[ i ][ j-1 ], mat_theta_lim[ i ][ j ], len( self.cumul_S_fft ) ) )
                    self.cumul_S_fft.append( [ [ 0.0 for n in self.loop_bins_t ] for a in self.loop_channels ] )
                    self.event_counts.append( 0 )

            self.map_bins_part.append( ( vec_r_lim[ i ], vec_r_lim[ i+1 ], [] ) )

            # This is the original bin

            for j in range( len( mat_theta_lim[ i ] ) ):
                self.map_bins_part[ -1 ][ 2 ].append( ( mat_theta_lim[ i ][ j-1 ], mat_theta_lim[ i ][ j ], len( self.cumul_S_fft ) ) )
                self.cumul_S_fft.append( [ [ 0.0 for n in self.loop_bins_t ] for a in self.loop_channels ] )
                self.event_counts.append( 0 )

            if i < len( mat_theta_lim )-1:

                # This is the "top" sideband bin

                self.map_bins_part.append( ( vec_r_lim[ i+1 ], vec_r_lim[ i+2 ], [] ) )

                for j in range( len( mat_theta_lim[ i ] ) ):
                    self.map_bins_part[ -1 ][ 2 ].append( ( mat_theta_lim[ i ][ j-1 ], mat_theta_lim[ i ][ j ], len( self.cumul_S_fft ) ) )
                    self.cumul_S_fft.append( [ [ 0.0 for n in self.loop_bins_t ] for a in self.loop_channels ] )
                    self.event_counts.append( 0 )

        self.vector_distributions = [ vector_distribution() for i in range( len( self.cumul_S_fft )+1 ) ]

    def IncludeEvent( self, S ):
        S_fft = []
        amps = []

        dataS = np.asarray(S)
        dataS = np.sum(dataS, axis=0)
        peaks, properties = find_peaks(dataS.transpose(), prominence = 1, width=10)
        width_half = peak_widths(dataS.transpose(), peaks, rel_height=0.5)
        if (np.mean(dataS[0:3000])>np.mean(dataS[3000:10000]) and np.mean(dataS[0:3000])>np.mean(dataS[10000:15000]) and np.mean(dataS[0:3000])>np.mean(dataS[30000:32000])): return False
        if (np.mean(dataS[0:3000]) > 1.05*np.mean(dataS[30000:32000])) : return False 
        if (1.01*np.mean(dataS[0:3000]) < np.mean(dataS[30000:32000])) : return False

        if ( len(peaks) == 0 or len(peaks) > 1 or any(width_half)>2000) : return False

        #plt.plot(dataS.transpose())
        #plt.plot(peaks, dataS[peaks].transpose(), "xr")
        #plt.show()

        
        tmp_S = np.array(S)
        avg =  np.mean(tmp_S[:,2000:15000],axis=1) 
 
        for i in range(len(S)):
            S[i] = S[i]-avg[i]
        
        for a in self.loop_channels:
            S_fft.append( np.fft.fft( S[ a ] ).tolist() )
            #coefs_fft = [ ( 1.0-( self.J[ a ][ n ]/( np.absolute( S_fft[ -1 ][ n ] )**2 ) ) )*S_fft[ -1 ][ n ] for n in self.loop_bins_t ]
            #amps.append( float( sum( np.fft.ifft( coefs_fft ).tolist() ).real ) )

 
        #using scipy wiener filter
        noise_w = np.array([self.J[a][a] for a in range(len(S))])
        amps = [ sum( wiener( S[ a ], mysize=75, noise = noise_w[a].real ) ) for a in range(len(S)) ]

        E = sum( amps )

        if E < self.E_min or E > self.E_max:
            return False

        r = self.calc_r( amps )
        theta = get_angle_std( self.calc_theta( amps ) )

        found = False

        for i in range( len( self.map_bins_part ) ):
            if r > self.map_bins_part[ i ][ 0 ] and r < self.map_bins_part[ i ][ 1 ]:
                for j in range( len( self.map_bins_part[ i ][ 2 ] ) ):
                    if check_angle( theta, self.map_bins_part[ i ][ 2 ][ j ][ 0 ], self.map_bins_part[ i ][ 2 ][ j ][ 1 ] ):
                        k = self.map_bins_part[ i ][ 2 ][ j ][ 2 ]

                        for a in self.loop_channels:
                            for n in self.loop_bins_t:
                                self.cumul_S_fft[ k ][ a ][ n ] += S_fft[ a ][ n ]

                        self.event_counts[ k ] += 1
                        found = True

                        self.vector_distributions[ k ].add( theta, r )

        self.vector_distributions[ -1 ].add( theta, r )

        return found

    def GetTemplates( self ):
        templates = []

        for i in range( len( self.cumul_S_fft ) ):
            if self.event_counts[ i ] > 0:
                templates.append( [] )

                for a in self.loop_channels:
                    templates[ -1 ].append( [ float( coef.real ) for coef in np.fft.ifft( self.cumul_S_fft[ i ][ a ] ).tolist() ] )

                    for n in self.loop_bins_t:
                        templates[ -1 ][ -1 ][ n ] /= float( self.event_counts[ i ] )

        return templates

    def GetMapBinsPart( self ):
        map_bins_part = []
        template_count = 0

        for i in range( len( self.map_bins_part ) ):
            map_bins_part.append( ( self.map_bins_part[ i ][ 0 ],
                                    self.map_bins_part[ i ][ 1 ],
                                    []                           ) )

            for j in range( len( self.map_bins_part[ i ][ 2 ] ) ):
                if self.event_counts[ self.map_bins_part[ i ][ 2 ][ j ][ 2 ] ] > 0:
                    map_bins_part[ -1 ][ 2 ].append( ( self.map_bins_part[ i ][ 2 ][ j ][ 0 ],
                                                       self.map_bins_part[ i ][ 2 ][ j ][ 1 ],
                                                       template_count                         ) )

                    template_count += 1

                else:

                    # If a bin of the partition space does not has statistics to build the template, then label it with a -1

                    map_bins_part[ -1 ][ 2 ].append( ( self.map_bins_part[ i ][ 2 ][ j ][ 0 ],
                                                       self.map_bins_part[ i ][ 2 ][ j ][ 1 ],
                                                       -1                                     ) )

        return map_bins_part

    def Draw( self, path ):
        graphs = []

        for vd in self.vector_distributions:
            if vd.get_size() > 0:
                graphs.append( TGraph( vd.get_size(), vd.get_array_x(), vd.get_array_y() ) )

            else:
                graphs.append( TGraph() )

        limits_x = [ -math.pi                    , math.pi                       ]
        limits_y = [ self.map_bins_part[ 0 ][ 0 ], self.map_bins_part[ -1 ][ 1 ] ]

        lines = []

        for i in range( len( self.map_bins_part ) ):
            if i%3 == 0:
                r_bot = self.map_bins_part[ i ][ 0 ]
                r_top = self.map_bins_part[ i ][ 1 ]

                for j in range( len( self.map_bins_part[ i ][ 2 ] ) ):
                    theta_clockw = self.map_bins_part[ i ][ 2 ][ j ][ 0 ]
                    theta_anticw = self.map_bins_part[ i ][ 2 ][ j ][ 1 ]

                    lines.append( TLine( theta_clockw, r_bot, theta_clockw, r_top ) )
                    lines.append( TLine( theta_anticw, r_bot, theta_anticw, r_top ) )

                    if theta_anticw > theta_clockw:
                        lines.append( TLine( theta_clockw, r_bot, theta_anticw, r_bot ) )
                        lines.append( TLine( theta_clockw, r_top, theta_anticw, r_top ) )

                    else:
                        lines.append( TLine( theta_clockw, r_bot, math.pi     , r_bot ) )
                        lines.append( TLine( -math.pi    , r_bot, theta_anticw, r_bot ) )
                        lines.append( TLine( theta_clockw, r_top, math.pi     , r_top ) )
                        lines.append( TLine( -math.pi    , r_top, theta_anticw, r_top ) )

        for line in lines:
            line.SetLineColor( 15 )

        for i in range( len( self.map_bins_part ) ):
            if i%3 == 0:
                for j in range( len( self.map_bins_part[ i ][ 2 ] ) ):
                    graphs_subset = [ graphs[ -1 ], graphs[ self.map_bins_part[ i ][ 2 ][ j ][ 2 ] ] ]
                    colors = [ 3, 2 ]

                    if len( self.map_bins_part[ i ][ 2 ] ) > 1:
                        graphs_subset.append( graphs[ self.map_bins_part[ i ][ 2 ][ j-1 ][ 2 ] ] )
                        colors.append( 4 )

                        if len( self.map_bins_part[ i ][ 2 ] ) > 2:
                            graphs_subset.append( graphs[ self.map_bins_part[ i ][ 2 ][ ( j+1 )%len( self.map_bins_part[ i ][ 2 ] ) ][ 2 ] ] )
                            colors.append( 4 )

                    if i > 0:
                        graphs_subset.append( graphs[ self.map_bins_part[ i-1 ][ 2 ][ j ][ 2 ] ] )
                        colors.append( 4 )

                    if i < len( self.map_bins_part )-1:
                        graphs_subset.append( graphs[ self.map_bins_part[ i+1 ][ 2 ][ j ][ 2 ] ] )
                        colors.append( 4 )

                    filename = path+'/'+self.__class__.__name__+'_map_'+str( ( i/3 )+1 )+'_'+str( j+1 )+'.png'
                    draw_graphs( graphs_subset, colors, 0.5, '#theta', 'r', limits_x, limits_y, filename, lines )

        for graph in graphs:
            graph.Delete()

        for vd in self.vector_distributions:
            vd.reset()

        templates = self.GetTemplates()

        hists = []

        for a in self.loop_channels:
            hists.append( TH1F( create_name_hist(), '', self.num_bins_t, 0.0, float( self.num_bins_t ) ) )

        map_bins_part = self.GetMapBinsPart()

        for i in range( len( map_bins_part ) ):
            if i%3 == 0:
                for j in range( len( map_bins_part[ i ][ 2 ] ) ):
                    k = map_bins_part[ i ][ 2 ][ j ][ 2 ]

                    if k > -1:
                        for a in self.loop_channels:
                            for n in self.loop_bins_t:
                                hists[ a ].SetBinContent( n+1, templates[ k ][ a ][ n ] )

                        filename = path+'/'+self.__class__.__name__+'_pulse_'+str( ( i/3 )+1 )+'_'+str( j+1 )+'.png'
                        draw_hists( hists, [ 2+a for a in self.loop_channels ], 'Time (bin)', '', filename, '', legend = True)
                    
                        for hist in hists:
                            hist.Reset()

                    if i > 0:
                        k = map_bins_part[ i-1 ][ 2 ][ j ][ 2 ]

                        if k > -1:
                            for a in self.loop_channels:
                                for n in self.loop_bins_t:
                                    hists[ a ].SetBinContent( n+1, templates[ k ][ a ][ n ] )

                            filename = path+'/'+self.__class__.__name__+'_pulse_'+str( ( i/3 )+1 )+'_'+str( j+1 )+'_bot.png'
                            #draw_hists( hists, [ 2+a for a in self.loop_channels ], 'Time (bin)', '', filename, '')
                    
                            for hist in hists:
                                hist.Reset()

                    if i < len( map_bins_part )-1:
                        k = map_bins_part[ i+1 ][ 2 ][ j ][ 2 ]

                        if k > -1:
                            for a in self.loop_channels:
                                for n in self.loop_bins_t:
                                    hists[ a ].SetBinContent( n+1, templates[ k ][ a ][ n ] )

                            filename = path+'/'+self.__class__.__name__+'_pulse_'+str( ( i/3 )+1 )+'_'+str( j+1 )+'_top.png'
                            #draw_hists( hists, [ 2+a for a in self.loop_channels ], 'Time (bin)', '', filename, '')

                            for hist in hists:
                                hist.Reset()

        for hist in hists:
            hist.Delete()

        return True

File: cpp_version/TemplateGeneratorNxM.cc
#include "header.h"
#include "TemplateGeneratorNxM.h"

#include <iostream>
#include <cmath>
#include <stdio.h>
#include <sstream>    
#include <cstddef>
#include <string>
#include <complex>
#include <vector>

#include "TGraph.h"
#include "TH1F.h"
#include "TLine.h"





TemplateGeneratorNxM::TemplateGeneratorNxM(std::vector<std::vector<std::vector<std::complex<double>>>> V, double E_MIN, double E_MAX, std::vector<double> vec_r_lim, std::vector<std::vector<double>> mat_theta_lim) : len_t_lim(mat_theta_lim.size()), len_r_lim(vec_r_lim.size()), num_channels(V.size()), num_bins_t(V[0][0].size()), E_min(E_MIN), E_max(E_MAX) {

    initialize(J, num_channels, num_bins_t);


    for (int a = 0; a < num_channels; a++){
        for (int n = 0; n < num_bins_t; n++){
            J[a][n] = V[a][a][n];               //auto-power  spectral densities (diagonals of cross-power spectral density matrix)
        }
    }


    std::vector<std::vector<std::complex<double>>> cumul_S_fft_0;
    initialize(cumul_S_fft_0, num_channels, num_bins_t);

    for (int a = 0; a < num_channels; a++){
        for (int n = 0; n < num_bins_t; n++){
            cumul_S_fft_0[a][n] = 0.0;        //a 2D vector of zeros that will be appended for each template, placeholder for all events included in a template's production
        }
    }


    //initializing cumul_S_fft and map_bins_part, described in the header file of this class
    int len_t_lim_i;
    for (int i = 0; i < len_t_lim; i++){
        len_t_lim_i = mat_theta_lim[i].size();
                
        if (i>0){

            std::vector<std::vector<std::vector<double>>> temp { {{vec_r_lim[i-1]}}, {{vec_r_lim[i]}}, {} };

            map_bins_part.push_back(temp);


            for (int j = 0; j< len_t_lim_i; j++){
                if(j==0){
                    map_bins_part[map_bins_part.size()-1][2].push_back( std::vector<double> {mat_theta_lim[i][len_t_lim_i-1], mat_theta_lim[i][j], (double)cumul_S_fft.size()} );
                }else{
                    map_bins_part[map_bins_part.size()-1][2].push_back( std::vector<double> {mat_theta_lim[i][j-1], mat_theta_lim[i][j], (double)cumul_S_fft.size()} );
                }
                cumul_S_fft.push_back(cumul_S_fft_0);
                event_counts.push_back(0);
            }

        }

        std::vector<std::vector<std::vector<double>>> temp { {{vec_r_lim[i]}}, {{vec_r_lim[i+1]}}, {} };
        map_bins_part.push_back(temp);

        for (int j = 0; j< len_t_lim_i; j++){
            if(j==0){
                map_bins_part[map_bins_part.size()-1][2].push_back( std::vector<double> {mat_theta_lim[i][len_t_lim_i-1], mat_theta_lim[i][j], (double)cumul_S_fft.size()} );
            }else{
                map_bins_part[map_bins_part.size()-1][2].push_back( std::vector<double> {mat_theta_lim[i][j-1], mat_theta_lim[i][j], (double)cumul_S_fft.size()} );
            }
            cumul_S_fft.push_back(cumul_S_fft_0);
            event_counts.push_back(0);
        }

        if (i<len_t_lim-1){

            std::vector<std::vector<std::vector<double>>> temp { {{vec_r_lim[i+1]}}, {{vec_r_lim[i+2]}}, {} };

            map_bins_part.push_back(temp);

            for (int j = 0; j< len_t_lim_i; j++){
                if(j==0){
                    map_bins_part[map_bins_part.size()-1][2].push_back( std::vector<double> {mat_theta_lim[i][len_t_lim_i-1], mat_theta_lim[i][j], (double)cumul_S_fft.size()} );
                }else{
                    map_bins_part[map_bins_part.size()-1][2].push_back( std::vector<double> {mat_theta_lim[i][j-1], mat_theta_lim[i][j], (double)cumul_S_fft.size()} );
                }
                cumul_S_fft.push_back(cumul_S_fft_0);
                event_counts.push_back(0);
            }

        }
    }

    vector_distributions = std::vector<vector_distribution>(cumul_S_fft.size()+1);      

    for (int i = 0; i<cumul_S_fft.size()+1; i++){
        vector_distribution Vect;
        vector_distributions[i] = Vect;
    }
}


//function to include an event in the creation of templates (we include all events in the "calibration sample")
bool TemplateGeneratorNxM::IncludeEvent(std::vector<std::vector<std::complex<double>>> S){
    std::vector<std::vector<std::complex<double>>> S_fft;
    initialize(S_fft, num_channels, num_bins_t);
    std::vector<double> amps(num_channels);

    std::vector<std::complex<double>> coefs_fft(num_bins_t);
    std::vector<std::complex<double>> IFFT(num_bins_t);

    for(int a = 0; a < num_channels; a++){

        amps[a] = 0.0;

        fft(S_fft[a], S[a]);

        for (int n = 0; n < num_bins_t; n++){  
            coefs_fft[n] = ( 1.0-( J[ a ][ n ]/( std::conj(S_fft[a][n])*(S_fft[a][n]) ) ) )*S_fft[ a ][ n ];
        } 
        ifft(IFFT, coefs_fft);

        for (int n = 0; n < num_bins_t; n++){
            amps[a] += std::real(IFFT[n]);
        }
    }


    //pre-construction of the quantities we want to reconstruct
    double E = 0.0;

    for (int a = 0; a < num_channels; a++){
        E += amps[a];
    }
    
    if (E<E_min || E>E_max){
        return false;
    }else{

        double r;
        double theta;
        calc_r(r, amps);
        calc_theta(theta, amps);
        get_angle_std(theta, theta);

        bool found = false;

        //searching for all bins (user-defined or otherwise) that the included event lies within
        for(int i = 0; i < map_bins_part.size(); i++){
            if (r > map_bins_part[i][0][0][0] && r < map_bins_part[i][1][0][0]){
                for(int j = 0; j < map_bins_part[i][2].size(); j++){
                    bool truf = false;
                    check_angle(truf, theta, map_bins_part[i][2][j][0], map_bins_part[i][2][j][1]);
                    if (truf == true){
                        int k = std::round(map_bins_part[i][2][j][2]);
                        for (int a = 0; a < num_channels; a++){
                            for (int n = 0; n < num_bins_t; n++){
                                cumul_S_fft[k][a][n] += S_fft[a][n];   //including event in average of traces (templates)
                            }
                        }
                        event_counts[k] += 1;
                        found = true;

                        vector_distributions[k].add(theta,r);           //discussed in header file of this class
                    }
                }
            }
        }

        vector_distributions[vector_distributions.size()-1].add(theta,r);
        return found;


    }

}


//get generated templates
void TemplateGeneratorNxM::GetTemplates(std::vector<std::vector<std::vector<std::complex<double>>>> &ans){

    std::vector<std::complex<double>> IFFT(num_bins_t);
    std::vector<std::vector<std::complex<double>>> TEMPLATE;
    initialize(TEMPLATE, num_channels, num_bins_t);

    for (int i = 0; i < cumul_S_fft.size(); i++){
        if (event_counts[i] > 0){
            for(int a = 0; a < num_channels; a++){

                ifft(IFFT, cumul_S_fft[i][a]); 

                for(int n = 0; n < num_bins_t; n++){
                    TEMPLATE[a][n] = {std::real(IFFT[n])/((double)event_counts[i]), 0.0};
                }
	        }
            ans.push_back(TEMPLATE);   //only save templates which have events in them   
            
        }
    }
}




//get map_bins_part that corresponds to the selected templates in the function above.
void TemplateGeneratorNxM::GetMapBinsPart(std::vector<std::vector<std::vector<std::vector<double>>>> &map_bins_p){

    int template_count = 0;

    for (int i = 0; i<map_bins_part.size(); i++){


        std::vector<std::vector<std::vector<double>>> temp { {{map_bins_part[i][0][0][0]}}, {{map_bins_part[i][1][0][0]}}, {} };
        map_bins_p.push_back(temp);

        for (int j = 0; j<map_bins_part[i][2].size(); j++){
            int k = std::round(map_bins_part[i][2][j][2]);
            if (event_counts[k] > 0){
                map_bins_p[map_bins_p.size()-1][2].push_back( std::vector<double> {map_bins_part[i][2][j][0], map_bins_part[i][2][j][1], (double)template_count} );
                template_count += 1;
            }else{
                map_bins_p[map_bins_p.size()-1][2].push_back( std::vector<double> {map_bins_part[i][2][j][0], map_bins_part[i][2][j][1], -1} );
            }
        }
    }
}



//Draw calibration set in (r,theta) space, and bins for all possible optimal filters. Also draw some templates.
void TemplateGeneratorNxM::Draw(std::string path){

    std::string filename;

    const double pi = 3.14159265358979323846;

    int vds = vector_distributions.size();
    int maps_size = map_bins_part.size();

    std::vector<TGraph*> graphs(vds);

    for (int i = 0; i<vds; i++){
        if (vector_distributions[i].get_size() > 0){
            int n = vector_distributions[i].get_size();
            double * array_x = new double[n];
            double * array_y = new double[n];
            vector_distributions[i].get_array_x(array_x, n);
            vector_distributions[i].get_array_y(array_y, n);
            graphs[i] = new TGraph( n, array_x, array_y );
            delete[] array_x;
            delete[] array_y;
        }else{
            graphs[i] = new TGraph();
        }
    }

    std::vector<double> limits_x {-pi, pi};
    std::vector<double> limits_y {map_bins_part[0][0][0][0], map_bins_part[map_bins_part.size()-1][1][0][0]};
	
    std::vector<TLine*> lines;

    double r_bot, r_top, theta_clockw, theta_anticw;
    int maps_size_i2;

    for (int i = 0; i<maps_size; i++){
        if (i%3 == 0){
            maps_size_i2 = map_bins_part[i][2].size();

            r_bot = map_bins_part[i][0][0][0];
            r_top = map_bins_part[i][1][0][0];

            for (int j = 0; j<maps_size_i2; j++){
                theta_clockw = map_bins_part[i][2][j][0];
                theta_anticw = map_bins_part[i][2][j][1];

                lines.push_back( new TLine(theta_clockw, r_bot, theta_clockw, r_top));
                lines.push_back( new TLine(theta_anticw, r_bot, theta_anticw, r_top));

                if(theta_anticw > theta_clockw){
                    lines.push_back( new TLine( theta_clockw, r_bot, theta_anticw, r_bot ));
                    lines.push_back( new TLine( theta_clockw, r_top, theta_anticw, r_top ));
                }else{
                    lines.push_back( new TLine( theta_clockw, r_bot, pi     , r_bot ));
                    lines.push_back( new TLine( -pi    , r_bot, theta_anticw, r_bot ));
                    lines.push_back( new TLine( theta_clockw, r_top, pi     , r_top ));
                    lines.push_back( new TLine( -pi    , r_top, theta_anticw, r_top ));
                }
            }
            
        }
    }

    for (int i = 0; i<lines.size(); i++){
        lines[i]->SetLineColor(15);
    }

    //num graphs = vds
    for (int i = 0; i<maps_size; i++){
        if(i%3 == 0){
            maps_size_i2 = map_bins_part[i][2].size();

            for (int j = 0; j<maps_size_i2; j++){
                std::vector<TGraph*> graphs_subset {graphs[vds-1], graphs[std::round(map_bins_part[i][2][j][2])]};
                std::vector<int> colors {3,2};

                if(maps_size_i2>1){
                    if(j==0){
                        graphs_subset.push_back(graphs[std::round(map_bins_part[i][2][maps_size_i2-1][2])]);
                        colors.push_back(4);
                    }else{
                        graphs_subset.push_back(graphs[std::round(map_bins_part[i][2][j-1][2])]);
                        colors.push_back(4);
                    }

                    if(maps_size_i2>2){
                        graphs_subset.push_back(graphs[std::round(map_bins_part[i][2][(j+1)%maps_size_i2][2])]);
                        colors.push_back(4);
                    }
                }

                if(i>0){
                    graphs_subset.push_back(graphs[std::round(map_bins_part[i-1][2][j][2])]);
                    colors.push_back(4);
                }

                if(i<(maps_size-1)){
                    graphs_subset.push_back(graphs[std::round(map_bins_part[i+1][2][j][2])]);
                    colors.push_back(4);
                }

		filename.clear();
                filename = path+"/TemplateGeneratorNxM"+"_map_"+std::to_string( ( i/3 )+1 )+'_'+std::to_string( j+1 )+".png" ;
                draw_graphs(graphs_subset, colors, 0.5, "#theta", "r", limits_x, limits_y, filename, lines); 
            }



        }
    }

    for (int ii = 0; ii<graphs.size(); ii++){
        graphs[ii]->Delete();
    }

    for (int ii = 0; ii<lines.size(); ii++){
        delete lines[ii];
    }

    for (int i = 0; i<vds; i++){
        vector_distributions[i].reset();
    }


    //histograms:

    std::vector<std::vector<std::vector<std::complex<double>>>> templates;
    GetTemplates(templates);

    std::vector<TH1F*> hists(num_channels);

    std::vector<int> A(num_channels);

    for(int a = 0; a < num_channels; a++){
        hists[a] = new TH1F(create_name_hist(), "", num_bins_t, 0.0, (float)num_bins_t);
        A[a] = 2+a;
    }

    std::vector<std::vector<std::vector<std::vector<double>>>> mbpg; //map_bins_part_get;
    GetMapBinsPart(mbpg);
    int mbpgs = mbpg.size();

    std::string filename1;
    std::string filename2;
    std::string filename3;

    int k;
    for (int i = 0; i<mbpgs; i++){
        if(i%3 == 0){
            maps_size_i2 = mbpg[i][2].size();
            for(int j = 0; j<maps_size_i2; j++){
                k = std::round(mbpg[i][2][j][2]);

                if (k > -1){
                    for (int a = 0; a < num_channels; a++){
                        for(int n = 0; n < num_bins_t; n++){
                            hists[a]->SetBinContent(n+1, std::real(templates[k][a][n]));
                        }
                    }

		            filename1.clear();
                    filename1 = path+"/TemplateGeneratorNxM"+"_pulse_"+std::to_string( ( i/3 )+1 )+"_"+std::to_string( j+1 )+".png";
                    draw_hists(hists, A, "Time_(bin)", "", filename1);

                    for (int ii = 0; ii<hists.size(); ii++){
                        hists[ii]->Reset();
                    }



                }

                if(i > 0){
                    k = std::round(mbpg[i-1][2][j][2]);
                    if(k > -1){
                        for (int a = 0; a < num_channels; a++){
                            for(int n = 0; n < num_bins_t; n++){
                                hists[a]->SetBinContent(n+1, std::real(templates[k][a][n]));
                            }
                        }

			            filename2.clear();
                        filename2 = path+"/TemplateGeneratorNxM"+"_pulse_"+std::to_string( ( i/3 )+1 )+"_"+std::to_string( j+1 )+"_bot.png";
                        draw_hists(hists, A, "Time_(bin)", "", filename2);

                        for (int ii = 0; ii<hists.size(); ii++){
                            hists[ii]->Reset();
                        }

                    }
                }


                if(i < (mbpgs-1)){
                    k = std::round(mbpg[i+1][2][j][2]);
                    if(k > -1){
                        for (int a = 0; a < num_channels; a++){
                            for(int n = 0; n < num_bins_t; n++){
                                hists[a]->SetBinContent(n+1, std::real(templates[k][a][n]));
                            }
                        }
			            filename3.clear();
                        filename3 = path+"/TemplateGeneratorNxM"+"_pulse_"+std::to_string( ( i/3 )+1 )+"_"+std::to_string( j+1 )+"_top.png";
                        draw_hists(hists, A, "Time_(bin)", "", filename3);

                        for (int ii = 0; ii<hists.size(); ii++){
                            hists[ii]->Reset();
                        }

                    }


                }
            }
        }
    }

    for (int ii = 0; ii < hists.size(); ii++){
        hists[ii]->Delete();
    }    

}


